### DISCUSSION TITLE
ROM Info.txt Request

### DISCUSSION THREAD

**Discussion by majorpbx** [sanni/cartreader#211] [Ideas & Suggestions]
First, this Cart Reader is awesome! Second, I'm wondering if it's possible to have the Cart Reader write an Info.txt file with the details of the rom and initially found hash values? Say something like below saved to the ROM destination folder.

`
##### GAMENAME - SYSTEM #####
Name: XXXXXXXXXXXXXX
ID: XXXX
Size: XXXXX
Save: XXXXXXXXXXX
Version: X.X
Saved To: N64/ROM/ROMNAME/XXX
File Type: ByteSwapped
CRC: XXXXXXXX
Checksum match/not match
Done (XXXs)
`

If No-Intro DAT's could be implemented you could also add a few other things like below.

`

DAT File Used: XXXXXXXXXXXXXXXXXXXX.dat
Matched Game: XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX/Did Not Match!
Verified: No (#PLEASESUBMITTONOINTRO)
CRC: XXXXXXXX
MD5: XXXXXXXXXXXXXXXXXXXXXXXXXXXX
SHA-1: XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
`

Thanks

**Comment by sanni** [comment]
Seems like a great first project for you to try. You could write it something like this:

````
// Open file on sd card
if (!myFile.open("info.txt", O_RDWR | O_CREAT)) {
  print_Error(F("SD Error"), true);
}

//Write the info
myFile.print(F("Name: "));
myFile.println(romName);

myFile.print(F("ID: "));
myFile.println(cartID);

myFile.print(F("Size: "));
myFile.print(cartSize);

myFile.println(F("MB"));
myFile.print(F("Save: "));

switch (saveType) {
  case 1:
    myFile.println(F("Sram"));
    break;
  case 4:
    myFile.println(F("Flashram"));
    break;
  case 5:
    myFile.println(F("4K Eeprom"));
    break;
  case 6:
    myFile.println(F("16K Eeprom"));
    break;
  default:
   myFile.println(F("unknown"));
    break;
 }

myFile.print(F("Version: 1."));
myFile.println(romVersion);

myFile.print(F("Saved To: "));
myFile.println(filePath);
myFile.print(F("CRC: "));
myFile.println(crcStr);

// Close the file:
myFile.close();
````

And put it there:

![hrerghrgegegege](https://user-images.githubusercontent.com/19169494/127721183-561c8f4d-f124-4e8e-a7f1-a82dfaa652fa.jpg)

**Reply by majorpbx** [reply]
Thank you for the idea and sample code. :)

**Reply by sanni** [reply]
I have implemented your idea it in the newest commit. It now saves a n64log.txt to N64/ROMS/ that keeps track of all dumps.

![zhergwrwte](https://user-images.githubusercontent.com/19169494/127982387-aee16b51-a893-4fa9-9dff-8929595455e0.jpg)

**Reply by majorpbx** [reply]
Oh nice! I've added code to check for the option to enable logging so that this can be used across all modules. The example output I have when testing is below. My code also handles if the user redumps due to an error. Attempting to add dumping the carts CRC1 and CRC2 as well.

![2021-08-05_11-36-48](https://user-images.githubusercontent.com/18235905/128378838-0747478c-722e-4a04-8d82-cac30004d884.jpg)

**Reply by majorpbx** [reply]
Much better now!

![2021-08-05_12-34-45](https://user-images.githubusercontent.com/18235905/128387471-8954a9fb-db09-4359-bbb5-9863c0d4b22a.jpg)

**Reply by sanni** [reply]
Nice 👍

**Reply by majorpbx** [reply]
Would you mind if I did a pull request with my changes?

**Reply by sanni** [reply]
Sure go ahead.

**Reply by mauruco** [reply]
I don't know if it's possible, but is there a possibility to add one more line to the log? I imagined something like "Original: true/false", to check if the cartridge was clone or original.

**Reply by majorpbx** [reply]
From what I understand you'd probably know if not for the various shell items that can tell you the cart is not original. Another thing you might find you have to dump it with forced settings or repro settings. Point is you should probably already know but that's just me.

**Reply by mauruco** [reply]
Sometimes I buy games in bad conditions, where only the inside is original and the rest has been replaced. I'll always open it to check, but it would be nice to have a second level of checking.
A check for the lazy.
I'll give you an example, I don't know if that's how it works. Let's say the cartridge is a reproduction "AM29F032", if the software already knows it is or can detect it, it would be nice to have a warning.
I'm from a developing country, there are many reproductions here, any warning is better.

**Reply by mauruco** [reply]
A second example would be.
I go to a store to buy a game, the seller tells me it's an original cartridge, the seller put a seal over the screw so I can't open it, because he won't get the cartridge back if I break the seal.
In this situation, I pull my Sanni Cartridge Reader into the store and am able to test whether or not it is genuine. :)

**Reply by majorpbx** [reply]
I get what your saying and totally understand. I would be suprised to see a store owner let you plug it in to a foreign device without making you buy it afterwards! After all how do they know your not wiping it to be mean? By all means I'm not saying you would do that just seems to me store owners I know would never let me willing do that without making me buy it with a no buy back option.

So for your example, if lets say you were able to connect it up. If you choose the "Game Cartridge" option and it dumps successfully without issue (by that I mean you don't have to tell the Cart Reader the ROM size, etc...) I'd say you have at least a legit cart board inside the shell.

**Reply by mauruco** [reply]
You're right. Furthermore, even if it is possible for one party, there will always be others that manual inspection will be inevitable.

**Comment by mariomadproductions** [comment]
If this text file could be created for all systems, and included sanni version info (hardware + firmware) and the current date, that would be really helpful for no-intro submissions, as it means we don't have to ask the dumper to provide that info themselves. It'd probably be best if each ROM file had an equivalent log file. So you'd have whatever.rom and then whatever.rom.dumpinfo.txt, so its easy to copy it when the ROM is copied.

**Reply by majorpbx** [reply]
It would for sure, but keep in mind current date/time is extremely limited without an RTC to work with. Meaning every time you turn the SCR off/on the time goes back to "12/31/2020 11:00 PM" or whatever your device defaults to.

Separate logs are possible and can with code updates be used for all roms instead of just N64.

**Reply by mariomadproductions** [reply]
Ah I see. Well date is less important, but having the hardware+firmware version of the sanni with each dump would be pretty useful on its own.

**Reply by majorpbx** [reply]
I'll look into it, I have a few edits to make and I'll add at least hardware+firmware to the logs as that's easy enough.

**Reply by mariomadproductions** [reply]
Great, thanks!

**Reply by majorpbx** [reply]
New sample log, still a single log for all N64 dumps but it's a start.

Name	: MarioGolf64
ID	        : NMFJ
ROM CRC1: d48944d1
ROM CRC2: b0d93a0e
Size	        : 24MB
Save	        : Sram
Version	: 1.0
Saved To: N64/ROM/MarioGolf64/130
Dumped	: 16MAY2022 09:14:51PM
HW Ver	: 3
FW Ver	: 8.3
CRC         : 911f179a [Match]
Time	: 142

**Reply by mariomadproductions** [reply]
Looks good. I guess if the RTC is not installed, the dump date can just say `[Not available]` or something like that.

**Reply by mariomadproductions** [reply]
I think another feature that would be useful would be to print out what cart type was detected so that the user can make sure that if a cart uses a certain PCB configuration/mapper, it is detected and dumped appropriately for that configuration.

**Reply by majorpbx** [reply]
I am totally onboard with an update that encompasses that feature and have been thinking about it for a while now. However, the reality of the situation is that each cart that this supports has different read methods, power requirements, clock settings, etc... While I'd love nothing more to see this feature implemented at this time it's simply not feasible for more than just the cart requirements. It would take a complete overhaul of the code, probably 32/64bit logic board, more RAM, etc...

I'm not saying it can't be done, just as you can imagine it will take a considerable amount of work. That and this project was created on the bases of community added updates. That fact alone means that there would be very few people able to have the skills required to keep up with coding of that nature.

**Reply by mariomadproductions** [reply]
I think maybe you misunderstood what I said, I was just suggesting that for instance the NES dumper would print the mapper its dumping the cart as to the log. What were you thinking of, I'm curious?

**Reply by majorpbx** [reply]
Ah gotcha, yep misread that's on me. I was and have been thinking to have the reader identify the type of cart once inserted. 😅 

However, your instance would be fine once and if logging is setup for all carts.

**Comment by sanni** [comment]
I added a global log file option that just duplicates everything it shows on screen(minus the menu/file explorer output) to a text file on the SD.

![oscr_log](https://user-images.githubusercontent.com/19169494/173229925-de55d681-0e0c-4dea-b78e-7d7889f7b645.jpg)

Not quite what you have asked for but I it was a quick solution since it only needed a few lines of code added. 

Also added tamanegi_taro's PCE database and ROM renaming function to the Gameboy code, just for fun.

![IMG_1328](https://user-images.githubusercontent.com/19169494/173230091-41521261-217f-4cc2-9f11-6e4079ac7589.jpg)

**Reply by mariomadproductions** [reply]
Thanks, that looks good.

**Reply by majorpbx** [reply]
Nice!

**Reply by sanni** [reply]
Added hardware and software version to the log and removed the "Press Button..." and "Select File" spam and also added the selected menu choices.

![affaafafaffa](https://user-images.githubusercontent.com/19169494/173589252-5f3cff05-8579-43fa-aecf-716b6697f802.jpg)

**Reply by mariomadproductions** [reply]
Thanks. If possible, could you put the logs for each cart in their own txt file beside the dump? And also add the date and time the dump was made?

**Comment by sanni** [comment]
Another update. Now part of the OSCR_LOG.txt gets copied to the dump directory after reading a ROM.

![IMG_1336](https://user-images.githubusercontent.com/19169494/174078270-f7b6d9f4-ee8c-4f41-b73d-1605a4543c7a.jpg)

![IMG_1338](https://user-images.githubusercontent.com/19169494/174078274-fac78fe1-dda2-4073-a021-0bdcca79daeb.jpg)
